### PR TITLE
Projections

### PR DESCRIPTION
**Pull request by prolic**
New Features:

- EventStore Queries
- EventStore Projections (using LinkTo & Emit)
- EventStore ReadModel Projections (updating your custom read model)
- added EventStore::delete(StreamName $streamName) method

Check the tests for how to use this. Docs are coming later.

Please check the latest commit https://github.com/prooph/event-store/pull/213/commits/3395617ee04401ea96f82b58ee354460f3c470a2:

Here I changed from stdClass to array for state. I am not 100% sure which one is better.

### CODE REVIEW

File: src/StreamProjection/AbstractStreamProjection.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    public function __construct(EventStore $eventStore, string $name, int $persistBatches, bool $enableEmit)
+    {
+        parent::__construct($eventStore);
+
+        $this->name = $name;
+        $this->persistBatches = $persistBatches;
+        $this->emitEnabled = $enableEmit;
+
+        if ($enableEmit && ! $this->eventStore->hasStream(new StreamName($name))) {
+            $this->eventStore->create(new Stream(new StreamName("$name"), new ArrayIterator()));
```
Comment (lucascourot) [inline src/StreamProjection/AbstractStreamProjection.php]: why did you put double quotes here?

File: src/StreamProjection/AbstractStreamProjection.php [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        return $this->name;
+    }
+
+    public function emit(Message $event): void
+    {
+        if (! $this->emitEnabled) {
+            throw new RuntimeException('Emit is disabled');
+        }
+
+        $this->eventStore->appendTo(new StreamName("$this->name"), new ArrayIterator([$event]));
```
Comment (lucascourot) [inline src/StreamProjection/AbstractStreamProjection.php]: why did you put double quotes here?

File: src/StreamProjection/AbstractQuery.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    public function init(callable $callback): Query
+    {
+        if (null !== $this->initCallback) {
+            throw new RuntimeException('Projection already initialized');
+        }
+
+        $this->state = $callback();
+
+        if (! $this->state instanceof stdClass) {
+            throw new RuntimeException('init callback is expected to return an instance of stdClass');
```
Comment (lucascourot) [inline src/StreamProjection/AbstractQuery.php]: All messages should start with an uppercase IMHO

**Comment by prolic** [comment]
@codeliner ready for final review

**Comment by codeliner** [comment]
looks great. I need more time to play with it but the concept is clear and an amazing addition :+1: 
Going to merge as-is and if I found issues later we can discuss them in follow up PRs.

File: src/Projection/AbstractReadModelProjection.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        $this->readModelProjection = $readModelProjection;
+    }
+
+    public function readModelProjection(): ReadModelProjection
+    {
+        return $this->readModelProjection;
+    }
+
+    public function delete(bool $deleteEmittedEvents): void
```
Comment (prolic) [inline src/Projection/AbstractReadModelProjection.php]: This is part of the interface and will delete the projection. As the in memory adapter is not persistent, there is nothing more to do here. But for other event store implementations things are different.

**Pull request merged**